Adapt simulation to take into account electrical profiles and adapt endpoints

Castavo opened the pull request:
closes: #2778 
closes: #2779

flomonster commented on api/openapi.yaml:
```diff
@@ -802,6 +819,87 @@ paths:
+                            format: float
+                          end:
+                            type: number
+                            format: float
```
Can you add:

- Description to the objects
- Example values

------

You can also create a component `ElectricalProfile` and reference it.

flomonster commented on api/osrd_infra/migrations/0018_add_electrical_profiles_to_schedules.py:
```diff
@@ -0,0 +1,47 @@
+                null=True,
+                on_delete=django.db.models.deletion.CASCADE,
+                to="osrd_infra.electricalprofilesset",
+            ),
```
Foreign key migration should be done using `run_sql_add_foreign_key_infra`. Examples of use can be found in the first migration.

flomonster commented on api/osrd_infra/urls.py:
```diff
@@ -13,6 +14,7 @@
 
 router = DefaultRouter(trailing_slash=True)
 router.register("infra", InfraView, basename="infra")
+router.register("electrical_profile_set", ElectricalProfileSetView, basename="electrical_profile_set")
```
```suggestion
router.register("electrical_profiles", ElectricalProfilesView, basename="electrical_profiles")
```

flomonster commented on api/osrd_infra/models/external_generated_inputs.py:
```diff
@@ -1,12 +1,13 @@
 from django.contrib.gis.db import models
```
Can you rename this file into `electrical_profile.py`

flomonster commented on api/osrd_infra/models/external_generated_inputs.py:
```diff
@@ -1,12 +1,13 @@
 class ElectricalProfilesSet(models.Model):
     name = models.CharField(max_length=128)
     data = models.JSONField(validators=[PydanticValidator(ElectricalProfilesList)])
+    level_order = models.JSONField(validators=[PydanticValidator(ElectricalProfileLevelOrder)], default=dict)
```
Can you include this data into the electrical profile schema.

flomonster commented on api/osrd_infra/views/electrical_profiles.py:
```diff
@@ -0,0 +1,19 @@
+
+    @action(detail=True, methods=["get"])
+    def level_order(self, request, pk=None):
+        return Response(ElectricalProfilesSet.objects.get(pk=pk).level_order, content_type="application/json")
```
Move these endpoints to editoast

flomonster commented on core/src/main/java/fr/sncf/osrd/cli/StandaloneSimulationCommand.java:
```diff
@@ -84,7 +84,7 @@ public int run() {
 
             // Add allowance
-            var allowance = new RJSAllowance[] {
+            var allowance = new RJSAllowance[]{
```
```suggestion
            var allowance = new RJSAllowance[] {
```

flomonster commented on core/src/main/java/fr/sncf/osrd/external_generated_inputs/ElectricalProfileMapping.java:
```diff
@@ -0,0 +1,63 @@
+package fr.sncf.osrd.external_generated_inputs;
```
```suggestion
package fr.sncf.osrd.electrical_profiles;
```

flomonster commented on core/src/main/java/fr/sncf/osrd/envelope_sim/EnvelopeSimContext.java:
```diff
@@ -20,6 +21,8 @@ public EnvelopeSimContext(
-        this.tractiveEffortCurveMap = rollingStock.mapTractiveEffortCurves(path, comfort);
+        var curvesAndConditions = rollingStock.mapTractiveEffortCurves(path, comfort);
+        this.tractiveEffortCurveMap = curvesAndConditions.curves();
+        this.modeAndProfilesUsed = curvesAndConditions.conditions();
```
With the idea to keep a constructor as simple as possible (without computation).
Can you add a static method `build` that compute `curvesAndConditions` and then call the constructor with the `tractiveEffortCurveMap` and `modeAndProfileUsed`.

flomonster reviewed (changes requested):
Thanks for this PR!

multun commented on api/openapi.yaml:
```diff
@@ -504,6 +511,9 @@ paths:
                     type: string
                   infra:
                     type: number
+                  electrical_profile_set:
```
what's that?

multun commented on api/openapi.yaml:
```diff
@@ -996,6 +1010,29 @@ components:
             - type: object
               properties:
                 error: { type: string }
+        modes_and_profiles:
```
why is this exported to the front-end?

multun commented on api/osrd_infra/models/electrical_profiles.py:
```diff
@@ -1,12 +1,12 @@
+from osrd_infra.schemas.external_generated_inputs import ElectricalProfileSet as ElectricalProfileSetSchema
 
 
 class ElectricalProfilesSet(models.Model):
```
the description says "Profile" and the class name "Profiles"

multun commented on api/osrd_infra/models/timetable.py:
```diff
@@ -1,8 +1,9 @@
 
 class Timetable(models.Model):
     infra = models.ForeignKey(Infra, on_delete=models.CASCADE)
+    electrical_profile_set = models.ForeignKey(ElectricalProfilesSet, on_delete=models.CASCADE, null=True, blank=True)
```
is'nt it linked to the infra?

multun commented on api/osrd_infra/models/train_schedule.py:
```diff
@@ -22,3 +22,4 @@ class TrainScheduleModel(models.Model):
     comfort = models.CharField(
         max_length=8, choices=[(x.value, x.name) for x in ComfortType], default=ComfortType.STANDARD
     )
+    modes_and_profiles = models.JSONField(default=[])
```
we usually have a schema validator. having a list as a default also is not safe, please use default_factory

multun commented on core/openapi.yaml:
```diff
@@ -605,6 +748,9 @@ components:
           minItems: 1
           items:
             $ref: "#/components/schemas/TrainSchedule"
+        electrical_profile_set:
+          type: string
+          example: "ID"
```
why does the back need to return that?

multun commented on core/src/main/java/fr/sncf/osrd/api/MiddleWareInteraction.java:
```diff
@@ -0,0 +1,41 @@
+import okhttp3.Request;
+import okhttp3.Response;
+
+public abstract class MiddleWareInteraction {
```
how about just "APIClient"?

multun commented on core/src/main/java/fr/sncf/osrd/api/StandaloneSimulationEndpoint.java:
```diff
@@ -66,25 +71,37 @@ public Response act(Request req) throws
             var infra = infraManager.load(request.infra, request.expectedVersion, recorder);
 
+            // load electrical profile set
+            var electricalProfileMap = electricalProfileSetManager.getProfileMap(request.electricalProfileSet);
```
I think ElectricalProfileMap is a better name, couln't the class use it too?

multun commented on core/src/main/java/fr/sncf/osrd/envelope_sim/EnvelopePath.java:
```diff
@@ -1,36 +1,54 @@
+    /**
+     * A range map of modeAndProfile objects containing only modes
+     */
+    private final ImmutableRangeMap<Double, ModeAndProfile> catenaryModeMap;
```
RangeMaps aren't exactly fast, have you benchmarked the impact of your implementation?

multun commented on core/src/main/java/fr/sncf/osrd/envelope_sim/PhysicsPath.java:
```diff
@@ -18,6 +18,8 @@ public interface PhysicsPath {
+    /** The catenary modes on a given position of the path */
+    RangeMap<Double, ModeAndProfile> getModeAndProfileMap(String powerClass);
+
+    record ModeAndProfile(String mode, String profile) {}
```
what's the "mode"?

Castavo replied:
I'll elaborate in all doctstrings

multun commented on core/src/main/java/fr/sncf/osrd/envelope_sim/PhysicsRollingStock.java:
```diff
@@ -56,4 +56,8 @@ static double getMaxEffort(double speed, TractiveEffortPoint[] tractiveEffortCur
     }
+
+    record CurvesAndConditions(RangeMap<Double, TractiveEffortPoint[]> curves,
+                               RangeMap<Double, PhysicsPath.ModeAndProfile> conditions) {
```
I just don't get it, aren't conditions used to get the curves? is it just to debug / display?

Commit:
api + core + editoast: take electrical profiles into account

api: models: add profile order attribute to electrical profile sets

api: electrical profiles: add endpoint to retrieve level order

api: add endpoints and openAPI to use electrical profiles

core: api: load electrical profiles from api

core: api: add tests

core: api: create electrical profile mapping class

core: api: add power class to electrical profile mapping

core: infra-generator: add more power classes in electrical profiles

core: rolling-stock: adapt data model

core: rename catenary profiles to catenary modes

core: electrical profiles: improve cache management

core: simulation: take electrical profiles into account

core: test: add electrical profile mapping tests

core: tests: add sim test with electrical profiles

core: railjson: move rjselectricalprofile.java

core: gather modes and electrical profiles used by train

core: rolling stock: add get curve on path test

core: add electrical profiles used to the response

api: add electrical profiles seen to simulation result

core: envelope_path: merge rangemap ranges with same values

move electrical profile endpoints from api to editoast

include level order inside electrical profile set schema

editoast + core: update electrical profile exchange format

editoast: add electrical profile tests

core: make envelope_sim_context init simple again

core: rename middleware_interaction to api_client

api: change electrical profile set foreign key migration

electrical profiles: improve docs

github-actions: update tarpaulin

editoast: move electrical profile sets queryables to view

core: change electrical profile result format

flomonster commented on api/osrd_infra/migrations/__init__.py:
```diff
@@ -2,10 +2,10 @@
     return migrations.RunSQL(
         f"""ALTER TABLE osrd_infra_{model_name}
-                ADD {field_name}_id INTEGER,
+                ADD {field_name}_id INTEGER {"NULL" * nullable},
```
Using a ternary condition looks better.

```suggestion
                ADD {field_name}_id INTEGER {"NULL" if nullable else ""},
```

flomonster commented on api/osrd_infra/models/timetable.py:
```diff
@@ -1,8 +1,9 @@
 
 class Timetable(models.Model):
     infra = models.ForeignKey(Infra, on_delete=models.CASCADE)
+    electrical_profile_set = models.ForeignKey(ElectricalProfileSet, on_delete=models.CASCADE, null=True, blank=True)
```
`blank` doesn't make sens since it's not a `CharField`.

flomonster commented on core/examples/generated/lib/railjson_generator/external_generated_inputs.py:
```diff
@@ -37,6 +36,9 @@ def save(self, path):
-            __root__=[profile.to_rjs() for profile in self.electrical_profiles]
+        return external_generated_inputs.ElectricalProfileSet(
+            levels=[profile.to_rjs() for profile in self.electrical_profiles],
+            level_order={
+                "25000": ["25000", "22500", "20000"],
+            },
```
Is this normal ? It should be part of the schema no?

flomonster commented on editoast/openapi.yaml:
```diff
@@ -1009,3 +1095,17 @@ components:
+        track_ranges:
+          type: array
+          items:
+            $ref: "#/components/schemas/TrackRange"
```
Missing line feed

flomonster commented on editoast/src/schema/electrical_profiles.rs:
```diff
@@ -0,0 +1,200 @@
+            .first(conn)
+        {
+            Ok(ep_set) => Ok(ep_set),
+            Err(DieselError::NotFound) => {
+                Err(Box::new(ElectricalProfilesApiError::NotFound(ep_set_id)))
+            }
+            Err(e) => Err(Box::new(ElectricalProfilesApiError::DieselError(e))),
+        }
+    }
+
+    pub fn retrieve_data(
+        conn: &mut PgConnection,
+        ep_set_id: i64,
+    ) -> Result<ElectricalProfileSet, Box<dyn ApiError>> {
+        let ep_set_wrapper = Self::retrieve(conn, ep_set_id);
+        match ep_set_wrapper {
+            Ok(ep_set_wrapper) => {
+                match serde_json::from_value::<ElectricalProfileSet>(ep_set_wrapper.data) {
+                    Ok(ep_set) => Ok(ep_set),
+                    Err(e) => Err(Box::new(ElectricalProfilesApiError::InternalError(
+                        e.to_string(),
+                    ))),
+                }
+            }
+            Err(e) => Err(e),
+        }
+    }
+
+    pub fn list(
+        conn: &mut PgConnection,
+    ) -> Result<Vec<ElectricalProfileSetMetaData>, Box<dyn ApiError>> {
+        match dsl::osrd_infra_electricalprofileset
+            .select((dsl::id, dsl::name))
+            .load(conn)
+        {
+            Ok(ep_sets) => Ok(ep_sets),
+            Err(e) => Err(Box::new(ElectricalProfilesApiError::DieselError(e))),
+        }
+    }
+}
```
Queryable, errors and implementations should be moved to the view. 
The `schema` module is here to declare serializers.

flomonster commented on editoast/src/views/electrical_profiles.rs:
```diff
@@ -0,0 +1,74 @@
+    .await
+}
+
+pub fn routes() -> Vec<Route> {
+    routes![list, get, get_level_order]
+}
```
Can you move this at the top of the file.

flomonster reviewed (changes requested):
Almost there!
